### PR TITLE
Kiali 812 - Check if route rule versions labels has pods running

### PR DESCRIPTION
**Pull request by xeviknal**
Given a route rule routing traffic to v1 and v2, add a new validation if v1 or v2 doesn't have any pod running.

### CODE REVIEW

File: services/business/istio_validations.go [hunk 1]
```diff
@@ -51,8 +51,14 @@ func (in *IstioValidationsService) GetNamespaceValidations(namespace string) (mo
...
 
+	pods, err := in.k8s.GetNamespacePods(namespace)
+	if err != nil {
+		log.Warningf("Cannot get pods for namespace %v.", namespace)
+		return nil, err
+	}
+
 	objectCheckers := []ObjectChecker{
-		checkers.RouteRuleChecker{istioDetails.RouteRules},
+		checkers.RouteRuleChecker{Namespace: namespace, PodList: pods.Items, RouteRules: istioDetails.RouteRules},
```
Comment (lucasponce) [inline services/business/istio_validations.go]: I am not sure if this is correct.
For the GetServiceValidations, the pods are filtered by the service, or it should be if I read it correctly.
For the namespace, you are fetching all pods and this is checking against versions without matching per labels.
So, if I am not wrong we might have a bad-pod-1234 with labels version=v1, version=v2 and matches correctly.

The idea to add the RouteRuleChecker was to enable validations that are independent of the service, to not replicate in Istio Config what we are showing in Service Details.

Then, I would leave in GetNamespaceValidations validations that are per namespace, but not attached to the services, I think.

Perhaps, that means to split the RouteRuleChecker, or if PodList is not provided, to not use the pod checker on this.

Or, if we want to maintain the consistency, the pods should be narrowed/filtered per service in the evaluation.

If I didn't miss anything in the code, I couldn't review it entirely yet.

Comment (xeviknal) [reply services/business/istio_validations.go]: > For the GetServiceValidations, the pods are filtered by the service, or it should be if I read it correctly.

You are right, they are filtered by app label (service level).

> For the namespace, you are fetching all pods and this is checking against versions without matching per labels.

You are right, for the GetNamespaceValidations I am not filtering per service. This solution then it could trigger some false positive. 

> The idea to add the RouteRuleChecker was to enable validations that are independent of the service, to not replicate in Istio Config what we are showing in Service Details.

I don't know if I fully follow you. I thought that we wanted to show same validations for each item within IstioConfig view. If we are using RouteRuleChecker, then it means that we will show weight/precedence validations as well.
IMO, it might be somewhat confusing to have some validations in one page but not to the other. 

> Or, if we want to maintain the consistency, the pods should be narrowed/filtered per service in the evaluation.

IMO, I would go this way. But perhaps I am missing something.

File: services/business/checkers/route_rules/version_presence_checker.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	return podFound
+}
+
+func getSelector(routeRule kubernetes.IstioObject, rawLabels interface{}) labels.Selector {
+	routeLabels := map[string]string{}
+	for key, value := range rawLabels.(map[string]interface{}) {
+		routeLabels[key] = value.(string)
+	}
+
+	routeLabels[config.Get().ServiceFilterLabelName] = getServiceName(routeRule)
```
Comment (xeviknal) [inline services/business/checkers/route_rules/version_presence_checker.go]: @lucasponce now it is filtering by service label. This way, it is removed the false positive we talked before.

File: kubernetes/kubernetes_service.go [hunk 3]
```diff
@@ -105,6 +105,12 @@ func (in *IstioClient) GetPods(namespace string, labelsSet labels.Set) (*v1.PodL
 	return in.k8s.CoreV1().Pods(namespace).List(meta_v1.ListOptions{LabelSelector: formatted})
 }
 
+// GetPods returns the pods definitions for a given namespace
```
Comment (lucasponce) [inline kubernetes/kubernetes_service.go]: Copy/paste typo, it should be "GetNamespacePods"

File: services/business/checkers/route_rule_checker.go [hunk 4]
```diff
@@ -4,11 +4,14 @@ import (
 	"github.com/kiali/kiali/kubernetes"
 	"github.com/kiali/kiali/services/business/checkers/route_rules"
 	"github.com/kiali/kiali/services/models"
+	"k8s.io/api/core/v1"
```
Comment (lucasponce) [inline services/business/checkers/route_rule_checker.go]: Bad import order, it should be
standard imports
LINE
third pary
LINE
kiali imports

If not @jmazzitelli will chase you :-)

File: services/business/checkers/route_rule_checker.go [hunk 5]
```diff
@@ -42,15 +45,16 @@ func (in RouteRuleChecker) runGroupChecks() models.IstioValidations {
 }
 
 // enabledCheckersFor returns the list of all individual enabled checkers.
-func enabledCheckersFor(object kubernetes.IstioObject) []Checker {
+func (in RouteRuleChecker) enabledCheckersFor(object kubernetes.IstioObject) []Checker {
```
Comment (lucasponce) [inline services/business/checkers/route_rule_checker.go]: I am not sure if I like to extract functions for one-line assign.
Indeed I am adding this lines without the enabledCheckersFor, sorry, I guess a comment is enough in assignment :-).

File: services/business/checkers/route_rules/version_presence_checker.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	valid := true
+	validations := make([]*models.IstioCheck, 0)
+
+	slice := reflect.ValueOf(checker.RouteRule.GetSpec()["route"])
+	if slice.Kind() != reflect.Slice {
+		return validations, valid
+	}
+
+	for i := 0; i < slice.Len(); i++ {
+		route := slice.Index(i).Interface().(map[string]interface{})
```
Comment (lucasponce) [inline services/business/checkers/route_rules/version_presence_checker.go]: I think this cast should be protected, as a typo in the routerule yaml is possible and you might not cast this, so you would get then a nice panic.
A simple route, ok := ... should help.

File: services/business/checkers/route_rule_checker.go (hunk 4 above)
Comment (xeviknal) [reply services/business/checkers/route_rule_checker.go]: Upsie! Just forgot that rule here.

File: services/business/checkers/route_rules/version_presence_checker.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	}
+
+	return validations, valid
+}
+
+func (checker VersionPresenceChecker) hasMatchingPod(selector labels.Selector) bool {
+	podFound := false
+
+	for _, pod := range checker.PodList {
+		podFound = podFound || selector.Matches(labels.Set(pod.ObjectMeta.Labels))
```
Comment (lucasponce) [inline services/business/checkers/route_rules/version_presence_checker.go]: podFound = podFound || selector.Matches(labels.Set(pod.ObjectMeta.Labels)) ?

Not better just 

podFound = selector.Matches(labels.Set(pod.ObjectMeta.Labels))

As you break if true below ?

File: services/business/checkers/route_rules/version_presence_checker.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+			break
+		}
+	}
+
+	return podFound
+}
+
+func getSelector(routeRule kubernetes.IstioObject, rawLabels interface{}) labels.Selector {
+	routeLabels := map[string]string{}
+	for key, value := range rawLabels.(map[string]interface{}) {
```
Comment (lucasponce) [inline services/business/checkers/route_rules/version_presence_checker.go]: Cast should be protected, as this comes from some yaml with potential malformed types.

File: services/business/checkers/route_rules/version_presence_checker.go (hunk 2 above)
Comment (lucasponce) [reply services/business/checkers/route_rules/version_presence_checker.go]: It is good, as we know app x version label should exist on pods, so, this looks better now.
Thanks.

File: services/business/checkers/route_rules/version_presence_checker.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+		routeLabels[key] = value.(string)
+	}
+
+	routeLabels[config.Get().ServiceFilterLabelName] = getServiceName(routeRule)
+	return labels.Set(routeLabels).AsSelector()
+}
+
+func getServiceName(routeRule kubernetes.IstioObject) string {
+	serviceName := ""
+	if destination, ok := routeRule.GetSpec()["destination"]; ok {
```
Comment (lucasponce) [inline services/business/checkers/route_rules/version_presence_checker.go]: This assumes Spec exists. This could not pass some test with nil objects. I started to add them as I introduced a bug without notice for a similar condition like this, so I recommend to check it out first.

File: services/business/checkers/route_rules/version_presence_checker_test.go [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+	"github.com/stretchr/testify/assert"
+	"k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/kubernetes"
+)
+
+func TestCheckerWithPodsMatching(t *testing.T) {
```
Comment (lucasponce) [inline services/business/checkers/route_rules/version_presence_checker_test.go]: I suggest to add some basic test with inputs with nil, or empty objects to ensure things don't crash.

File: services/business/checkers/route_rules/version_presence_checker_test.go [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+	assert := assert.New(t)
+	conf := config.NewConfig()
+	config.Set(conf)
+
+	// Setup mocks
+	podList := []v1.Pod{
+		fakePodsForLabels(map[string]string{"app": "reviews", "version": "v1", "stage": "production"}),
```
Comment (lucasponce) [inline services/business/checkers/route_rules/version_presence_checker_test.go]: Just to double check, a map[string]string it could be different than map[string]interface{}, so are you sure that k8s cast labels.Set as a map[string]string vs map[string]interface{} ?

If not in runtime it can skip some cast.

File: services/business/istio_validations.go (hunk 1 above)
Comment (lucasponce) [reply services/business/istio_validations.go]: No, it is good, I wanted to maintain consistency between both, that's the correct way, but I didn't want to pull all data related to services in the Istio Config logic.

I guess filtering pods ver labels "app x version" is what it is necessary.

I am also working actively in these files too and I was also thinking loud.

File: services/business/istio_validations_test.go [hunk 12]
```diff
@@ -371,3 +379,28 @@ func fakeCombinedServices(services []string) *kubernetes.ServiceList {
...
 	return &serviceList
 }
+
+func fakePods() *v1.PodList {
+	return &v1.PodList{
+		Items: []v1.Pod{
+			v1.Pod{
+				ObjectMeta: meta_v1.ObjectMeta{
+					Name: "reviews-12345-hello",
+					Labels: map[string]string{
```
Comment (lucasponce) [inline services/business/istio_validations_test.go]: Commented out, just to be sure that the mock pod has the correct type used by the real pod, map[string]string vs map[string]interface{}.

It just to double check this.

Comment (xeviknal) [reply services/business/istio_validations_test.go]: Yes, k8s uses map[string]string for labels whereas IstioObject uses map[string]interface{}. I had to deal with that before and had to add those castings instead.

File: services/business/checkers/route_rules/version_presence_checker_test.go (hunk 11 above)
Comment (xeviknal) [reply services/business/checkers/route_rules/version_presence_checker_test.go]: labels.Set is a map[string]string by definition. So no problem in casting.
Dealing with IstioObjects is where most of the pain is. The definition of the struct is basically map[string]interface{}

**Comment by xeviknal** [comment]
@lucasponce comments approached!

**Comment by xeviknal** [comment]
Thanks for the exhaustive review lucas!
